sympy/combinatorics/rewritingsystem.py

```diff
@@ -0,0 +1,222 @@
+from __future__ import print_function, division
+
+from sympy import S
+from sympy.combinatorics.free_groups import FreeGroupElement
+
+class RewritingSystem(object):
+    '''
+    A class implementing rewriting systems for `FpGroup`s.
+
+    References
+    ==========
+    [1] Epstein, D., Holt, D. and Rees, S. (1991).
+        The use of Knuth-Bendix methods to solve the word problem in automatic groups.
+        Journal of Symbolic Computation, 12(4-5), pp.397-414.
+
+    [2] GAP's Manual on its KBMAG package
+        https://www.gap-system.org/Manuals/pkg/kbmag-1.5.3/doc/manual.pdf
+
+    '''
+    def __init__(self, group):
+        self.group = group
+        self.alphabet = group.generators
+        self._is_confluent = None
+
+        # these values are taken from [2]
+        self.maxeqns = 32767 # max rules
+        self.tidyint = 100 # rules before tidying
+
+        # _max_exceeded is True if maxeqns is exceeded
+        # at any point
+        self._max_exceeded = False
+
+        # dictionary of reductions
+        self.rules = {}
+        self._init_rules()
+
+    def set_max(self, n):
+        '''
+        Set the maximum number of rules that can be defined
+
+        '''
+        if self._max_exceeded and n > self.maxeqns:
+            self._max_exceeded = False
+        self.maxeqns = n
+        return
+
+    @property
+    def is_confluent(self):
+        '''
+        Return `True` if the system is confluent
+
+        '''
+        if self._is_confluent is None:
+            self._is_confluent = self._check_confluence()
+        return self._is_confluent
+
+    def _init_rules(self):
+        identity = self.group.free_group.identity
+        for r in self.group.relators:
+            self.add_rule(r, identity)
+        self._remove_redundancies()
+        return
+
+    def add_rule(self, w1, w2):
+        new_keys = set()
+        if len(self.rules) + 1 > self.maxeqns:
+            self._is_confluent = self._check_confluence()
```

Comment:
Hmm, actually maybe `._check_confluence()` shouldn't try to add rules at all. Otherwise if someone runs it, it could return `False` while in practice the system is already confluent because of the rules it added.